### PR TITLE
fix(server): subscription memory leak

### LINKED ISSUE #6156: bug: memory leak in trpc 11 subscriptions.
### Provide environment information

Binaries:
  Node: 18.20.4 - /usr/local/bin/node
  Yarn: 1.22.19 - /usr/local/bin/yarn
  npm: 10.7.0 - /usr/local/bin/npm
  pnpm: 8.9.2 - /usr/local/bin/pnpm
  bun: 1.0.7 - /usr/local/bin/bun
npmPackages:
  @trpc/server: 11.0.0-rc.593 => 11.0.0-rc.593+f73cd3fd9 
  typescript: ^5.5.4 => 5.6.2 
  
node --version
v18.20.4

Nest app with typeorm that uses trpc.

### Describe the bug

When emitting large strings, either using observables or the yield function in generators, I notice memory usage increases while the WebSocket connection is open. Once the connection closes, the memory is released.

In my original setup, where I detected the leak, I had Redis subscribers listening for events within the observable, and I was properly freeing all resources and closing any open connections.

This issue occurs in version 11 of tRPC but not in version 10.

`node --inspect --expose_gc --max_old_space_size=700 --optimize_for_size dist/src/main.js`

Using the same setup for trpc 11 and 10:

// TRPC 11
The script uses approximately 198.88 MB - cache 0
The script uses approximately 198.21 MB - cache 0
The script uses approximately 198.93 MB - cache 0
The script uses approximately 196.57 MB - cache 0
The script uses approximately 195.12 MB - cache 0
The script uses approximately 195.05 MB - cache 0
The script uses approximately 195.04 MB - cache 0
The script uses approximately 195.11 MB - cache 0
The script uses approximately 195.22 MB - cache 0
The script uses approximately 195.6 MB - cache 0
The script uses approximately 195.29 MB - cache 0
The script uses approximately 195.89 MB - cache 0
The script uses approximately 195.37 MB - cache 0
The script uses approximately 195.27 MB - cache 0
The script uses approximately 195.14 MB - cache 0
The script uses approximately 195.11 MB - cache 0
The script uses approximately 195.37 MB - cache 0
The script uses approximately 195.11 MB - cache 0
//HERE STARTS THE TRPC SUBSCRIPTION
The script uses approximately 200.96 MB - cache 0.06345272064208984
The script uses approximately 241.62 MB - cache 0.3078422546386719
The script uses approximately 213.58 MB - cache 0.3078422546386719
The script uses approximately 263.7 MB - cache 0.3078422546386719
The script uses approximately 238.31 MB - cache 0.3078422546386719
The script uses approximately 288.44 MB - cache 0.3078422546386719
The script uses approximately 263.23 MB - cache 0.30820369720458984
The script uses approximately 313.07 MB - cache 0.30820369720458984
The script uses approximately 286.77 MB - cache 0.30820369720458984
The script uses approximately 336.95 MB - cache 0.30820369720458984
The script uses approximately 311.63 MB - cache 0.30820369720458984
The script uses approximately 361.76 MB - cache 0.30820369720458984
The script uses approximately 336.34 MB - cache 0.30820369720458984
The script uses approximately 386.75 MB - cache 0.30820369720458984
The script uses approximately 361.23 MB - cache 0.30820369720458984
The script uses approximately 411.57 MB - cache 0.30820369720458984
The script uses approximately 430.67 MB - cache 0.30820369720458984
The script uses approximately 397.54 MB - cache 0.30820369720458984
The script uses approximately 447.75 MB - cache 0.30820369720458984
The script uses approximately 461.06 MB - cache 0.30820369720458984
The script uses approximately 474.29 MB - cache 0.30820369720458984
The script uses approximately 487.8 MB - cache 0.30820369720458984
The script uses approximately 462.29 MB - cache 0.30820369720458984
The script uses approximately 473.89 MB - cache 0.30820369720458984
The script uses approximately 560.98 MB - cache 0.30820369720458984
The script uses approximately 535.57 MB - cache 0.30820369720458984
The script uses approximately 585.57 MB - cache 0.30820369720458984
The script uses approximately 560.33 MB - cache 0.30820369720458984
The script uses approximately 610.53 MB - cache 0.30820369720458984
The script uses approximately 585.13 MB - cache 0.30820369720458984
The script uses approximately 598.48 MB - cache 0.30820369720458984
The script uses approximately 573.07 MB - cache 0.30820369720458984
The script uses approximately 623.29 MB - cache 0.30820369720458984
The script uses approximately 597.75 MB - cache 0.30820369720458984
The script uses approximately 646.25 MB - cache 0.30820369720458984

//TRPC 10
The script uses approximately 197.55 MB - cache 0
The script uses approximately 197.59 MB - cache 0
//HERE STARTS THE TRPC SUBSCRIPTION
The script uses approximately 204.9 MB - cache 2.951091766357422
The script uses approximately 223.55 MB - cache 0
The script uses approximately 196.85 MB - cache 0
The script uses approximately 196.32 MB - cache 0
The script uses approximately 217.12 MB - cache 0.12111759185791016
The script uses approximately 217.19 MB - cache 0.12111759185791016
The script uses approximately 217.48 MB - cache 0.12111759185791016
The script uses approximately 217.05 MB - cache 0.12111759185791016
The script uses approximately 217.25 MB - cache 0.12111759185791016
The script uses approximately 219.32 MB - cache 0.12111759185791016
The script uses approximately 217.07 MB - cache 0.12111759185791016
The script uses approximately 217.17 MB - cache 0.12111759185791016
The script uses approximately 217.26 MB - cache 0.12111759185791016
The script uses approximately 217.31 MB - cache 0.12111759185791016
The script uses approximately 217.22 MB - cache 0.12111759185791016
The script uses approximately 219.24 MB - cache 0.12111759185791016
The script uses approximately 217.23 MB - cache 0.12111759185791016
The script uses approximately 217.1 MB - cache 0.12111759185791016
The script uses approximately 217.45 MB - cache 0.12111759185791016
The script uses approximately 217.33 MB - cache 0.12111759185791016
The script uses approximately 217.19 MB - cache 0.12111759185791016
The script uses approximately 217.2 MB - cache 0.12111759185791016
The script uses approximately 219.34 MB - cache 0.12111759185791016
The script uses approximately 217.35 MB - cache 0.12111759185791016
The script uses approximately 219.59 MB - cache 0.12111759185791016
The script uses approximately 217.27 MB - cache 0.12111759185791016
The script uses approximately 217.74 MB - cache 0.12111759185791016
The script uses approximately 217.21 MB - cache 0.12111759185791016
The script uses approximately 217.5 MB - cache 0.12111759185791016
The script uses approximately 217.22 MB - cache 0.12111759185791016
The script uses approximately 217.51 MB - cache 0.12111759185791016
The script uses approximately 217.61 MB - cache 0.12111759185791016
The script uses approximately 217.34 MB - cache 0.12111759185791016
The script uses approximately 217.6 MB - cache 0.12111759185791016
The script uses approximately 217.34 MB - cache 0.12111759185791016
The script uses approximately 217.33 MB - cache 0.12111759185791016
The script uses approximately 217.75 MB - cache 0.12111759185791016
The script uses approximately 217.3 MB - cache 0.12111759185791016
The script uses approximately 217.35 MB - cache 0.12111759185791016
The script uses approximately 217.39 MB - cache 0.12111759185791016
The script uses approximately 219.45 MB - cache 0.12111759185791016
The script uses approximately 217.66 MB - cache 0.12111759185791016
The script uses approximately 217.26 MB - cache 0.12111759185791016
The script uses approximately 217.35 MB - cache 0.12111759185791016
The script uses approximately 219.52 MB - cache 0.12111759185791016
The script uses approximately 217.28 MB - cache 0.12111759185791016
The script uses approximately 219.69 MB - cache 0.12111759185791016
The script uses approximately 217.38 MB - cache 0.12111759185791016
The script uses approximately 219.6 MB - cache 0.12111759185791016
The script uses approximately 217.67 MB - cache 0.12111759185791016
The script uses approximately 217.26 MB - cache 0.12111759185791016
The script uses approximately 217.41 MB - cache 0.12111759185791016
The script uses approximately 217.31 MB - cache 0.12111759185791016
The script uses approximately 217.55 MB - cache 0.12111759185791016
The script uses approximately 217.74 MB - cache 0.12111759185791016
The script uses approximately 217.28 MB - cache 0.12111759185791016
The script uses approximately 217.81 MB - cache 0.12111759185791016
The script uses approximately 217.4 MB - cache 0.12111759185791016
The script uses approximately 219.52 MB - cache 0.12111759185791016
The script uses approximately 217.36 MB - cache 0.12111759185791016
The script uses approximately 219.5 MB - cache 0.12111759185791016
The script uses approximately 217.42 MB - cache 0.12111759185791016
The script uses approximately 217.35 MB - cache 0.12111759185791016
The script uses approximately 217.39 MB - cache 0.12111759185791016
The script uses approximately 217.43 MB - cache 0.12111759185791016
The script uses approximately 219.58 MB - cache 0.12111759185791016
The script uses approximately 217.7 MB - cache 0.12111759185791016
The script uses approximately 217.42 MB - cache 0.12111759185791016
The script uses approximately 217.32 MB - cache 0.12111759185791016
The script uses approximately 217.41 MB - cache 0.12111759185791016
The script uses approximately 219.63 MB - cache 0.12111759185791016
The script uses approximately 217.39 MB - cache 0
The script uses approximately 217.15 MB - cache 0
The script uses approximately 217.15 MB - cache 0
The script uses approximately 217.05 MB - cache 0
The script uses approximately 217.09 MB - cache 0
The script uses approximately 217.44 MB - cache 0
The script uses approximately 217.08 MB - cache 0
The script uses approximately 217.08 MB - cache 0
The script uses approximately 219.54 MB - cache 0
The script uses approximately 217.14 MB - cache 0
The script uses approximately 219.36 MB - cache 0
The script uses approximately 217.16 MB - cache 0
The script uses approximately 219.37 MB - cache 0
The script uses approximately 217.29 MB - cache 0
The script uses approximately 219.53 MB - cache 0
The script uses approximately 217.39 MB - cache 0
The script uses approximately 217.11 MB - cache 0
The script uses approximately 217.12 MB - cache 0
The script uses approximately 217.19 MB - cache 0
The script uses approximately 219.41 MB - cache 0
The script uses approximately 217.36 MB - cache 0

### Link to reproduction

Not needed.

### To reproduce

```
...
 trpc.procedure.
...
 .subscription(
        async function* (this, _) {
          while (true) { //for example purposes.
            const bigString = Array(100000).fill('a').join('');
            yield bigString;
            await new Promise((resolve) => setTimeout(resolve, 100));
          }
            ....
          };
        }.bind(this),
      ),
```

### Additional information

I grabbed one of those "aaa..." strings in the memory inspector, and here’s what I found. Something related with the new getRawInput introduced in version 11 holding data?
image

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/6156">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/6156/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by oleasteo**
Closes #6156

### CODE REVIEW

File: packages/server/src/adapters/ws.ts [hunk 1]
```diff
@@ -332,6 +343,10 @@ export function getWSConnectionHandler<TRouter extends AnyRouter>(
               jsonrpc,
               result,
             });
+
+            // free up references for garbage collection
+            next = null;
```
Comment (KATT) [inline packages/server/src/adapters/ws.ts:348]: This is set on next iteration
```suggestion
```

File: packages/server/src/adapters/ws.ts [hunk 2]
```diff
@@ -313,7 +324,7 @@ export function getWSConnectionHandler<TRouter extends AnyRouter>(
               break;
             }
 
-            const result: TRPCResultMessage<unknown>['result'] = {
+            result = {
```
Comment (KATT) [inline packages/server/src/adapters/ws.ts:327]: This looks unnecessary - if it's in the loop instead it should be GC'd, no?

File: packages/server/src/adapters/ws.ts [hunk 3]
```diff
@@ -282,8 +283,18 @@ export function getWSConnectionHandler<TRouter extends AnyRouter>(
...
         run(async () => {
+          // We need those declarations outside the loop for garbage collection reasons. If they
+          // were declared inside, they would not be freed until the next value is present.
+          let next:
+            | null
+            | TRPCError
+            | Awaited<
+                typeof abortPromise | ReturnType<(typeof iterator)['next']>
+              >;
+          let result: null | TRPCResultMessage<unknown>['result'];
```
Comment (KATT) [inline packages/server/src/adapters/ws.ts:294]: I don't see how this is needed

File: packages/server/src/adapters/ws.ts (hunk 2 above)
Comment (KATT) [reply packages/server/src/adapters/ws.ts:327]: I see your comment _"We need those declarations outside the loop for garbage collection reasons."_ 


can you explain why?

it's within a closure... it should be freed automatically?

Comment (oleasteo) [reply packages/server/src/adapters/ws.ts:327]: Unfortunately, no. I thought so, too. Until I had those test cases proving me wrong.

I mentioned this in https://github.com/nodejs/node/issues/55585. It seems like a `const xyz = ...` inside a loop is not freed up at the end of the closure, but only when the same statement is reached again. I assume this to be some v8 optimization or bug. Maybe the node devs will tell me within that issue I opened on their side.

Comment (oleasteo) [reply packages/server/src/adapters/ws.ts:327]: Those cases are not as severe as the original reported issue, but still we'd have a dangling reference to the latest value. Seems to be a pretty easy error to have in nodejs, as even a `yield*` does have this issue itself xD

Comment (KATT) [reply packages/server/src/adapters/ws.ts:327]: oh wow, TIL!

File: packages/server/src/adapters/ws.ts (hunk 1 above)
Comment (KATT) [reply packages/server/src/adapters/ws.ts:348]: this one should be removable thoughhh

**Review by KATT — APPROVED** [review]

**Pull request merged**